packages/backend/src/modules/tvl2/AmountService.ts

```diff
@@ -0,0 +1,211 @@
+import { assert, Logger } from '@l2beat/backend-tools'
+import {
+  assertUnreachable,
+  EscrowEntry,
+  TotalSupplyEntry,
+  UnixTime,
+} from '@l2beat/shared-pure'
+import { Configuration } from '@l2beat/uif'
+import { partition } from 'lodash'
+
+import {
+  ERC20MulticallCodec,
+  NativeAssetMulticallCodec,
+} from '../../peripherals/multicall/codecs'
+import { MulticallClient } from '../../peripherals/multicall/MulticallClient'
+import {
+  MulticallRequest,
+  MulticallResponse,
+} from '../../peripherals/multicall/types'
+import { RpcClient } from '../../peripherals/rpcclient/RpcClient'
+import { AmountRecord } from './repositories/AmountRepository'
+
+export type AmountConfiguration = EscrowEntry | TotalSupplyEntry
+
+export interface AmountServiceDependencies {
+  readonly rpcClient: RpcClient
+  readonly multicallClient: MulticallClient
+  readonly erc20Codec: ERC20MulticallCodec
+  readonly nativeAssetCodec?: NativeAssetMulticallCodec
+  logger: Logger
+}
+
+interface Amount {
+  configurationId: string
+  amount: bigint
+}
+
+export class AmountService {
+  constructor(private readonly $: AmountServiceDependencies) {
+    this.$.logger = $.logger.for(this)
+  }
+
+  public async fetchAmounts(
+    configurations: Configuration<AmountConfiguration>[],
+    blockNumber: number,
+    timestamp: UnixTime,
+  ): Promise<AmountRecord[]> {
+    const nativeAssetCodecAtBlock = getNativeAssetCodecAtBlock(
+      this.$.nativeAssetCodec,
+      blockNumber,
+    )
+    const [forRpc, forMulticall] = partition(
+      configurations,
+      // isNotSupportedByMulticall checks that type is "escrow"
```

Comment:
this comment is probably not needed any more

Reply:
[91152b1](https://github.com/l2beat/l2beat/pull/3266/commits/91152b1b98d4c042d116ffc77cbc49031e2ed700)